Air-gapped EIB Deployments (#204)
suse-edge/suse-edge.github.io: asciidoc/guides/air-gapped-eib-deployments.adoc

:warning-caption: :warning:
endif::[]

== Intro

This guide will show how to deploy Rancher and NeuVector completely air-gapped on SLE Micro 5.5+ utilizing <<components-eib,Edge Image Builder(EIB)>>. With this, you'll be able to boot into a customized, ready to boot (CRB) image created by EIB and have Rancher and/or NeuVector deployed on a RKE2/K3s cluster without an internet connection or any manual steps. This configuration is highly desirable for customers that want to pre-bake all artifacts required for deployment into the CRB image so that the artifacts are immediately available on boot. This guide will explain configuring EIB and verifying the Rancher/NeuVector deployments.

It's important to note that the <<components-rancher,Rancher>> deployment that will be demonstrated will be highly slimmed down for demonstration purposes. For your actual deployments, additional artifacts may be necessary depending on your configuration. This guide will demonstrate the process of adding artifacts on an as-needed basis.

Further information about <<components-rancher,Rancher>> and <<components-neuvector,NeuVector>> can be found in the links to their respective guides.

== Prerequisites

> If you're following this guide, it's assumed that you are already familiar with <<components-eib,Edge Image Builder(EIB)>>, if not, please follow the EIB <<quickstart-eib,quick start guide>> to familiarize yourself with the process.

Author: This will probably need to change but putting it here for now.

> == Libvirt Network Configuration

Author: Let me know if this is ok. It operates under the assumption that users will know how to use this, and further, assumes users will manually set the mac address of the VMs later when deploying the CRB image.

Reviewer: I don't know that we need to tell them how to run it, just how to build it, right?

> To demo the air-gapped deployment, this guide will be done using a simulated air-gapped libvirt network and the following configuration will be tailored to that. For your own deployments, you may have to modify the `host1.local.yaml` that will be introduced in the next step.

Reviewer 2: I think this will be better formatted as a note. Thoughts?

```suggestion
> NOTE: To demo the air-gapped deployment, this guide will be done using a simulated air-gapped libvirt network and the following configuration will be tailored to that. For your own deployments, you may have to modify the `host1.local.yaml` configuration that will be introduced in the next step.
```

Reviewer: I'd ask that we use 

```
[NOTE]
====
blah
====
```

for consistency

If you would like to use the same libvirt network configuration, the steps for it are as follows:

[,console]
----
cat << EOF > isolatednetwork.xml
<network>
  <name>isolatednetwork</name>
  <bridge name='virbr1' stp='on' delay='0'/>
  <ip address='192.168.100.1' netmask='255.255.255.0'>
    <dhcp>
      <range start='192.168.100.2' end='192.168.100.254'/>
    </dhcp>
  </ip>
</network>
EOF
----

[,console]
----
virsh net-define isolatednetwork.xml
virsh net-start isolatednetwork
----

== Base EIB Directory Configuration

This guide features both Rancher and NeuVector deployments, but the base directory configuration for each will be the same, so we will set that up here.

> For simplicity, setting the variable `$EIB` to the full path of where you want the configuration directory is recommended, e.g. `EIB=/home/user/eib/config`.

Reviewer 2: Can we use `$CONFIG_DIR` instead? Just $EIB seems weird.

> For simplicity, setting the variable `$EIB` to the full path of where you want the configuration directory is recommended, e.g. `EIB=/home/user/eib/config`.

Reviewer 2: ```suggestion
For simplicity, setting the variable `$EIB` to the full path of where you want the configuration directory is recommended, e.g. `EIB=$HOME/eib`.
```

We will first create the necessary directories:
```
mkdir -p $EIB/base-images
mkdir -p $EIB/network
mkdir -p $EIB/kubernetes/helm/values
```

> Make sure to add whichever base image you plan to use into the `base-images` directory. This guide will focus on the Self Install ISO.

Reviewer 2: Can we add a link to the download page?

Reviewer: https://www.suse.com/download/sle-micro/

```
cp SLE-Micro.x86_64-5.5.0-Default-SelfInstall-GM2.install.iso $EIB/base-images/slemicro.iso

> ```

Reviewer 2: Should use the `[,shell]` syntax.

> Within the `network` directory, create a file named `host1.local.yaml` containing the following:

Reviewer 2: ```suggestion
Let's create a file containing the desired network configuration:
```

[,console]
----
cat << EOF > $EIB/network/host1.local.yaml
routes:
  config:
  - destination: 0.0.0.0/0
    metric: 100
    next-hop-address: 192.168.100.1
    next-hop-interface: eth0
    table-id: 254
  - destination: 192.168.100.0/24
    metric: 100
    next-hop-address:
    next-hop-interface: eth0
    table-id: 254
dns-resolver:
  config:
    server:
    - 192.168.100.1
    - 8.8.8.8
interfaces:
- name: eth0
  type: ethernet
  state: up
  mac-address: 34:8A:B1:4B:16:E7
  ipv4:
    address:
    - ip: 192.168.100.50
      prefix-length: 24
    dhcp: false
    enabled: true
  ipv6:
    enabled: false
EOF
----

> This gives the machine that we will deploy a static IP of `192.168.100.50` which will simplify the Rancher deployment.

Reviewer 2: The wording seems off here. Can we improve it?

For the base EIB image definition, we will use:

[,console]
----
cat << EOF > $EIB/eib-iso-definition.yaml
apiVersion: 1.0
image:
  imageType: iso
  arch: x86_64
  baseImage: slemicro.iso
  outputImageName: eib-image.iso
operatingSystem:
  users:
    - username: root
      encryptedPassword: \$6\$jHugJNNd3HElGsUZ\$eodjVe4te5ps44SVcWshdfWizrP.xAyd71CVEXazBJ/.v799/WRCBXxfYmunlBO2yp1hm/zb4r8EmnrrNCF.P/
      createHomeDir: true
kubernetes:

>   version: v1.28.7+rke2r1

Reviewer 2: ```suggestion
  version: v1.28.8+rke2r1
```

  network:
    apiVIP: 192.168.100.151
  nodes:
    - hostname: host1.suse.com

>       type: server

Reviewer 2: We can drop this section since this is a single node cluster.

EOF
----

[NOTE]
====
This creates a root user with the password `eib` that we can use to log in. Additionally, the `cat` command includes a backslash before each `$`, this is to make sure the command doesn't try to interpret `$`. If you were to copy this into the file directly, you must remove each `\` before each `$`.
====

The resulting file structure should now look like:
[,console]
----
├── eib-iso-definition.yaml
├── kubernetes/
│   └ helm/
│     └ values/
├── base-images/
│   └ slemicro.iso
└── network/  
    └ host1.local.yaml
----

== Rancher Installation Using Edge Image Builder

> While EIB typically completely air-gaps user manifests and Helm charts automatically, in the case where the user provided Helm charts/manifests side load images, manually defining the necessary images may be required. This is true for the case of Rancher.

Reviewer 2: ```suggestion
[WARNING]
====
EIB will parse and pre-download all images referenced in the provided Helm charts and Kubernetes manifests. However, some of those may be attempting to pull container images and create Kubernetes resources based on those at runtime. In these cases we have to manually specify the necessary images in the definition file if we want to set up a completely airgapped environment.
====
```

Let's begin by adding the Rancher Helm chart to the EIB image definition as well as the additional components it needs such as `cert-manager`:
[,console]

> ----

Reviewer 2: I'm definitely not a fan of having multiple versions of the definition file flying around. Can we have each section define its file just once?

apiVersion: 1.0
image:
  imageType: iso
  arch: x86_64
  baseImage: slemicro.iso
  outputImageName: eib-image.iso
operatingSystem:
  users:
    - username: root
      encryptedPassword: $6$jHugJNNd3HElGsUZ$eodjVe4te5ps44SVcWshdfWizrP.xAyd71CVEXazBJ/.v799/WRCBXxfYmunlBO2yp1hm/zb4r8EmnrrNCF.P/
      createHomeDir: true
kubernetes:

>   version: v1.28.7+rke2r1

Reviewer 2: ```suggestion
  version: v1.28.8+rke2r1
```

  network:
    apiVIP: 192.168.100.151
  nodes:
    - hostname: host1.suse.com

>       type: server

Reviewer 2: Not necessary.

  manifests:
    urls:
    - https://github.com/cert-manager/cert-manager/releases/download/v1.14.2/cert-manager.crds.yaml
  helm:
    charts:
      - name: rancher
        version: 2.8.3
        repositoryName: rancher-stable
        valuesFile: rancher-values.yaml
        targetNamespace: cattle-system
        createNamespace: true
        installationNamespace: kube-system
      - name: cert-manager
        installationNamespace: kube-system
        createNamespace: true
        repositoryName: jetstack
        targetNamespace: cert-manager
        version: 1.14.2
    repositories:
      - name: jetstack
        url: https://charts.jetstack.io

>       - name: rancher-stable

Reviewer 2: Shouldn't we use Rancher Prime?

Reviewer 3: Yes, we should

[,console]
----
cat << EOF > $EIB/kubernetes/helm/values/rancher-values.yaml
hostname: 192.168.100.50.sslip.io
replicas: 1
bootstrapPassword: "adminadminadmin"
systemDefaultRegistry: docker.io
useBundledSystemChart: true
EOF
----

[NOTE]
====
This sets the Rancher bootstrap password to `adminadminadmin`. Additionally, setting the `systemDefaultRegistry` to `docker.io` allows Rancher to automatically look for images in the embedded artifact registry started within the CRB image at boot.
====

To air-gap Rancher, we first need all of the images required by Rancher. For v2.8.3, this can be done by downloading the rancher-images.txt found in the https://github.com/rancher/rancher/releases/tag/v2.8.3[Rancher v2.8.3 release].

[,console]
----
wget https://github.com/rancher/rancher/releases/download/v2.8.3/rancher-images.txt
----

Within this file is every image used by Rancher for a full deployment. There are about 602 container images here which means that the resulting CRB image will be roughly 28GB+. In the following steps, we will significantly strip down the Rancher deployment. From there, you can add back any images you may need for your deployments.

[NOTE]
====

> The instructions for stripping down the Rancher deployment are highly influenced from the https://github.com/clemenko/rke_airgap_install/blob/main/hauler_all_the_things.sh[hauler air-gap all things script found here].

Reviewer 2: I don't think it makes sense to link to third party and unofficially supported sources. This is the official SUSE Edge 3.0 documentation so I'd propose to only list the necessary images without an explanation of the "trim down" process. Thoughts?

CC: @agracey

Reviewer: Agreed. I think it's fair to say that we use Hauler under the hoods but don't need to go into any more detail

    - name: rancher/pause:3.6
    - name: rancher/prom-alertmanager:v0.21.0
    - name: rancher/prometheus-auth:v0.2.2
    - name: rancher/prometheus-federator:v0.3.4
    - name: rancher/prom-node-exporter:v1.0.1
    - name: rancher/prom-prometheus:v2.18.2
    - name: rancher/pushprox-client:v0.1.0-rancher2-client
    - name: rancher/pushprox-proxy:v0.1.0-rancher2-proxy
    - name: rancher/rancher-agent:v2.8.3
    - name: rancher/rancher-csp-adapter:v3.0.1
    - name: rancher/rancher:v2.8.3
    - name: rancher/rancher-webhook:v0.4.3
    - name: rancher/rke2-cloud-provider:v1.28.2-build20231016
    - name: rancher/rke2-runtime:v1.28.8-rke2r1
    - name: rancher/rke2-upgrade:v1.28.8-rke2r1
    - name: rancher/rke-tools:v0.1.96
    - name: rancher/security-scan:v0.2.14
    - name: rancher/shell:v0.1.23
    - name: rancher/system-agent-installer-k3s:v1.28.8-k3s1
    - name: rancher/system-agent-installer-rke2:v1.28.8-rke2r1
    - name: rancher/system-agent:v0.3.6-suc
    - name: rancher/system-upgrade-controller:v0.13.1
    - name: rancher/ui-plugin-catalog:1.1.4
    - name: rancher/ui-plugin-operator:v0.1.1
    - name: rancher/webhook-receiver:v0.2.5

>     - name: rancher/kubectl:v1.20.2

Reviewer 3: kubectl 1.20?!

----

[NOTE]
====
From here, you can customize the deployment for your needs by appending any additional images you require to the `embeddedArtifactRegistry`.
====

> To build the image:

Author: Will need to change this to `registry.suse.com` but when I tried to pull it before I got an error so using this for now.

Reviewer 2: Can we discuss this online? We must use `registry.suse.com`.
